cli/command/telemetry.go

```diff
@@ -0,0 +1,214 @@
+package command
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/docker/distribution/uuid"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// TracerProvider is an extension of the trace.TracerProvider interface for CLI programs.
+type TracerProvider interface {
+	trace.TracerProvider
+	ForceFlush(ctx context.Context) error
+	Shutdown(ctx context.Context) error
+}
+
+// MeterProvider is an extension of the metric.MeterProvider interface for CLI programs.
+type MeterProvider interface {
+	metric.MeterProvider
+	ForceFlush(ctx context.Context) error
+	Shutdown(ctx context.Context) error
+}
+
+// TelemetryClient provides the methods for using OTEL tracing or metrics.
+type TelemetryClient interface {
+	// Resource returns the OTEL Resource configured with this TelemetryClient.
+	// This resource may be created lazily, but the resource should be the same
+	// each time this function is invoked.
+	Resource() *resource.Resource
+
+	// TracerProvider returns a TracerProvider. This TracerProvider will be configured
+	// with the default tracing components for a CLI program along with any options given
+	// for the SDK.
+	TracerProvider(ctx context.Context, opts ...sdktrace.TracerProviderOption) TracerProvider
+
+	// MeterProvider returns a MeterProvider. This MeterProvider will be configured
+	// with the default metric components for a CLI program along with any options given
+	// for the SDK.
+	MeterProvider(ctx context.Context, opts ...sdkmetric.Option) MeterProvider
+}
+
+func (cli *DockerCli) Resource() *resource.Resource {
+	return cli.res.Get()
+}
+
+func (cli *DockerCli) TracerProvider(ctx context.Context, opts ...sdktrace.TracerProviderOption) TracerProvider {
+	allOpts := make([]sdktrace.TracerProviderOption, 0, len(opts)+2)
+	allOpts = append(allOpts, sdktrace.WithResource(cli.Resource()))
+	if exp, err := dockerSpanExporter(ctx, cli); err != nil {
+		otel.Handle(err)
+	} else if exp != nil {
+		allOpts = append(allOpts, sdktrace.WithBatcher(exp))
+	}
+	allOpts = append(allOpts, opts...)
```

Comment:
```suggestion
	allOpts := append(opts, sdktrace.WithResource(cli.Resource()))
	if exp, err := dockerSpanExporter(ctx, cli); err != nil {
		otel.Handle(err)
	} else if exp != nil {
		allOpts = append(allOpts, sdktrace.WithBatcher(exp))
	}
	return sdktrace.NewTracerProvider(allOpts...)
```

Reply:
Maybe this isn't a concern, but the idea behind the original ordering is that user-provided options would occur later in the ordering than the default ones. For some options, in particular for resource, the last option will win and I didn't want an explicitly set resource to be overridden by something that was the default.

It's possible to override the default resource in a different way, but I thought it would be good to keep this pattern in case another default is needed in the future.

Reply:
It's not a huge concern to me, just a code nit :)
We can keep it as is